extensions/s3/src/main/java/io/deephaven/extensions/s3/S3SeekableByteChannel.java

```diff
@@ -0,0 +1,390 @@
+/**
+ * Copyright (c) 2016-2022 Deephaven Data Labs and Patent Pending
+ */
+package io.deephaven.extensions.s3;
+
+import io.deephaven.base.verify.Assert;
+
+import java.util.concurrent.CancellationException;
+
+import io.deephaven.util.channel.CachedChannelProvider;
+import io.deephaven.util.channel.SeekableChannelContext;
+import io.deephaven.util.SafeCloseable;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import software.amazon.awssdk.services.s3.S3AsyncClient;
+import software.amazon.awssdk.services.s3.S3Uri;
+import software.amazon.awssdk.services.s3.model.HeadObjectResponse;
+
+import java.io.IOException;
+import java.net.URI;
+import java.nio.ByteBuffer;
+import java.nio.channels.ClosedChannelException;
+import java.nio.channels.NonWritableChannelException;
+import java.nio.channels.SeekableByteChannel;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+
+/**
+ * {@link SeekableByteChannel} class used to fetch objects from AWS S3 buckets using an async client with the ability to
+ * read ahead and cache fragments of the object.
+ */
+final class S3SeekableByteChannel implements SeekableByteChannel, CachedChannelProvider.ContextHolder {
```

Comment:
Proposed patch for this file. Uses `CompletionStage` API to better manage the transformer's release, and cleans up a few warnings and looser-than-necessary access specifiers.
```
Index: extensions/s3/src/main/java/io/deephaven/extensions/s3/S3SeekableByteChannel.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/extensions/s3/src/main/java/io/deephaven/extensions/s3/S3SeekableByteChannel.java b/extensions/s3/src/main/java/io/deephaven/extensions/s3/S3SeekableByteChannel.java
--- a/extensions/s3/src/main/java/io/deephaven/extensions/s3/S3SeekableByteChannel.java	(revision e28b993feabeaaf3f594070b4826a4e9b51f4e4d)
+++ b/extensions/s3/src/main/java/io/deephaven/extensions/s3/S3SeekableByteChannel.java	(date 1706141007389)
@@ -5,7 +5,7 @@
 
 import io.deephaven.base.verify.Assert;
 
-import java.util.concurrent.CancellationException;
+import java.util.concurrent.*;
 
 import io.deephaven.util.channel.CachedChannelProvider;
 import io.deephaven.util.channel.SeekableChannelContext;
@@ -14,6 +14,7 @@
 import org.jetbrains.annotations.Nullable;
 import software.amazon.awssdk.services.s3.S3AsyncClient;
 import software.amazon.awssdk.services.s3.S3Uri;
+import software.amazon.awssdk.services.s3.model.GetObjectResponse;
 import software.amazon.awssdk.services.s3.model.HeadObjectResponse;
 
 import java.io.IOException;
@@ -22,10 +23,6 @@
 import java.nio.channels.ClosedChannelException;
 import java.nio.channels.NonWritableChannelException;
 import java.nio.channels.SeekableByteChannel;
-import java.util.concurrent.CompletableFuture;
-import java.util.concurrent.ExecutionException;
-import java.util.concurrent.TimeUnit;
-import java.util.concurrent.TimeoutException;
 
 
 /**
@@ -57,19 +54,11 @@
              * The future that will be completed with the fragment's bytes
              */
             @NotNull
-            private CompletableFuture<ByteBuffer> future;
+            private Future<ByteBuffer> future;
 
-            /**
-             * The {@link ByteBufferAsyncResponseTransformer} that will be used to complete the future
-             */
-            @NotNull
-            private ByteBufferAsyncResponseTransformer asyncResponseTransformer;
-
-            private FragmentState(final int fragmentIndex, @NotNull final CompletableFuture<ByteBuffer> future,
-                    @NotNull final ByteBufferAsyncResponseTransformer asyncResponseTransformer) {
+            private FragmentState(final int fragmentIndex, @NotNull final Future<ByteBuffer> future) {
                 this.fragmentIndex = fragmentIndex;
                 this.future = future;
-                this.asyncResponseTransformer = asyncResponseTransformer;
             }
         }
 
@@ -93,24 +82,21 @@
             return fragmentIndex % bufferCache.length;
         }
 
-        void setFragmentState(final int fragmentIndex, @NotNull final CompletableFuture<ByteBuffer> future,
-                @NotNull final ByteBufferAsyncResponseTransformer asyncResponseTransformer) {
+        private void setFragmentState(final int fragmentIndex, @NotNull final Future<ByteBuffer> future) {
             final int cacheIdx = getIndex(fragmentIndex);
             final FragmentState cachedEntry = bufferCache[cacheIdx];
             if (cachedEntry == null) {
-                bufferCache[cacheIdx] = new FragmentState(fragmentIndex, future, asyncResponseTransformer);
+                bufferCache[cacheIdx] = new FragmentState(fragmentIndex, future);
             } else {
                 // We should not cache an already cached fragment
                 Assert.neq(cachedEntry.fragmentIndex, "cachedEntry.fragmentIndex", fragmentIndex, "fragmentIdx");
 
                 // Cancel any outstanding requests for this cached fragment, and release the buffer for reuse
                 cachedEntry.future.cancel(true);
-                cachedEntry.asyncResponseTransformer.release();
 
                 // Reuse the existing entry
                 cachedEntry.fragmentIndex = fragmentIndex;
                 cachedEntry.future = future;
-                cachedEntry.asyncResponseTransformer = asyncResponseTransformer;
             }
         }
 
@@ -119,7 +105,7 @@
          * else will return {@code null}
          */
         @Nullable
-        CompletableFuture<ByteBuffer> getCachedFuture(final int fragmentIndex) {
+        private Future<ByteBuffer> getCachedFuture(final int fragmentIndex) {
             final FragmentState cachedFragment = bufferCache[getIndex(fragmentIndex)];
             if (cachedFragment != null && cachedFragment.fragmentIndex == fragmentIndex) {
                 return cachedFragment.future;
@@ -127,11 +113,11 @@
             return null;
         }
 
-        long getSize() {
+        private long getSize() {
             return size;
         }
 
-        void setSize(final long size) {
+        private void setSize(final long size) {
             this.size = size;
         }
 
@@ -139,9 +125,8 @@
         public void close() {
             // Cancel all outstanding requests
             for (final FragmentState fragmentState : bufferCache) {
-                if (fragmentState != null && fragmentState.future != null) {
+                if (fragmentState != null) {
                     fragmentState.future.cancel(true);
-                    fragmentState.asyncResponseTransformer.release();
                 }
             }
         }
@@ -219,8 +204,7 @@
             }
 
             // Wait till the current fragment is fetched
-            final CompletableFuture<ByteBuffer> currFragmentFuture =
-                    s3ChannelContext.getCachedFuture(currFragmentIndex);
+            final Future<ByteBuffer> currFragmentFuture = s3ChannelContext.getCachedFuture(currFragmentIndex);
             final ByteBuffer currentFragment;
             try {
                 currentFragment = currFragmentFuture.get(s3Instructions.readTimeout().toNanos(), TimeUnit.NANOSECONDS);
@@ -283,12 +267,12 @@
         final String range = "bytes=" + readFrom + "-" + readTo;
 
         final int numBytes = (int) (readTo - readFrom + 1);
-        final ByteBufferAsyncResponseTransformer asyncResponseTransformer =
-                new ByteBufferAsyncResponseTransformer(bufferPool.take(numBytes));
-        final CompletableFuture<ByteBuffer> future = s3AsyncClient.getObject(
-                builder -> builder.bucket(bucket).key(key).range(range),
-                asyncResponseTransformer);
-        s3ChannelContext.setFragmentState(fragmentIndex, future, asyncResponseTransformer);
+        final ByteBufferAsyncResponseTransformer<GetObjectResponse> asyncResponseTransformer =
+                new ByteBufferAsyncResponseTransformer<>(bufferPool.take(numBytes));
+        final CompletableFuture<ByteBuffer> future = s3AsyncClient
+                .getObject(builder -> builder.bucket(bucket).key(key).range(range), asyncResponseTransformer)
+                .whenComplete((response, throwable) -> asyncResponseTransformer.release());
+        s3ChannelContext.setFragmentState(fragmentIndex, future);
     }
 
     private IOException handleS3Exception(final Exception e, final String operationDescription) {
```

Reply:
Here are the unit tests I wrote when verifying my understanding of `CompletionStage`:
```
    @Test
    public void testCompletableFutureSuccess() throws ExecutionException, InterruptedException {
        final MutableBoolean called = new MutableBoolean();
        final CompletableFuture<Object> f1 = new CompletableFuture<>();
        final CompletableFuture<Object> f2 = f1
                .whenComplete((response, throwable) -> called.setValue(true));
        f1.complete("Hello");
        Assert.assertTrue(called.booleanValue());
        Assert.assertEquals("Hello", f2.get());
    }

    @Test
    public void testCompletableFutureError() {
        final MutableBoolean called = new MutableBoolean();
        final CompletableFuture<Object> f1 = new CompletableFuture<>();
        final CompletableFuture<Object> f2 = f1
                .whenComplete((response, throwable) -> called.setValue(true));
        f1.completeExceptionally(new RuntimeException("test"));
        Assert.assertTrue(called.booleanValue());
        try {
            f2.get();
            TestCase.fail("ExecutionException exception");
        } catch (ExecutionException expected) {
            //noinspection CallToPrintStackTrace
            expected.printStackTrace();
        } catch (InterruptedException e) {
            TestCase.fail("Unexpected InterruptedException");
        }
    }

    @Test
    public void testCompletableFutureCancel() {
        final MutableBoolean called = new MutableBoolean();
        final CompletableFuture<Object> f1 = new CompletableFuture<>();
        final CompletableFuture<Object> f2 = f1
                .whenComplete((response, throwable) -> called.setValue(true));
        f1.cancel(true);
        Assert.assertTrue(called.booleanValue());
        try {
            f2.get();
            TestCase.fail("ExecutionException exception");
        } catch (ExecutionException expected) {
            //noinspection CallToPrintStackTrace
            expected.printStackTrace();
        } catch (InterruptedException e) {
            TestCase.fail("Unexpected InterruptedException");
        }
    }
```